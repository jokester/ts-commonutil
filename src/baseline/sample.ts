import { defaultRng } from '../random/number';

export function sample<T>(input: readonly T[], rng = defaultRng): T {
  const index = Math.floor(input.length * rng());
  return input[index];
}

export function sampleSize<T>(from: readonly T[], count: number, rng = defaultRng): T[] {
  const reservoir: T[] = from.slice(0, Math.min(count, from.length));

  for (let i = count; i < from.length; i++) {
    const j = Math.floor(1 + rng() * i);
    if (j < count) {
      reservoir[j] = from[i];
    }
  }

  return reservoir;
}
