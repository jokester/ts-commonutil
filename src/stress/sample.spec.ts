import { sample, sampleSize } from './sample';

function fakeRng() {
  return 0.3;
}

describe('sample', () => {
  it('picks random value', () => {
    expect(sample([1, 2, 3], fakeRng)).toMatchSnapshot('sample([1,2,3])');
  });
  it('returns undefined on empty array', () => {
    expect(sample([])).toBeUndefined();
  });
});

describe('sampleSize', () => {
  it('return all elements when count==input.len', () => {
    expect(sampleSize([1, 2, 3], 3, fakeRng)).toEqual([1, 2, 3]);
  });
  it('return {count} elements when input is sufficient', () => {
    expect(sampleSize([1, 2, 3], 2, fakeRng)).toEqual([1, 3]);
  });
  it('return {input.length} elements when input is insufficient', () => {
    expect(sampleSize([2, 3], 3, fakeRng)).toEqual([2, 3]);
  });
});
