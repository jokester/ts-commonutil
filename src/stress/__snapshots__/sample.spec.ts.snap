// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sample picks random value: sample([1,2,3]) 1`] = `1`;

exports[`sampleSize picks specified size: sampleSize([1,2,3], 2) 1`] = `
[
  1,
  3,
]
`;
